### DISCUSSION TITLE
Adopt more error reporting features available in Clojure 1.10+

### DISCUSSION THREAD

**Discussion by daveliepmann** [clojure-emacs/cider#3338] [Ideas]
Since Clojure 1.10, Clojure provides ways to greatly improve error reporting in REPL tools. As today's standard-bearer for interactive development, CIDER should take full advantage of these new capabilities.

# Problem Statement
People consider error messages the top reason Clojure isn't more popular than it is, and the most important area of improvement for the Clojure experience. Year after year in the annual State of Clojure survey, the community ranks "error messages" a top priority for improvement and a major source of frustration:

 - [2022](https://www.surveymonkey.com/results/SM-QRiy0fSu3bmDK_2FSNMplVJw_3D_3D/), [2021](https://www.surveymonkey.com/results/SM-S2L8NR6K9/), [2018](https://www.surveymonkey.com/results/SM-9BC5FNJ68/), [2016](https://www.surveymonkey.com/results/SM-7K6NXJY3/), [2015](https://cognitect.com/blog/2016/1/28/state-of-clojure-2015-survey-results) - number 1 answer (often by a ~20% margin) to "What has been the biggest obstacle to using Clojure?" (or "What has been most frustrating or has prevented you from using Clojure more than you do now?"); also rated most important area of Clojure to improve
 - [2020](https://www.surveymonkey.com/results/SM-CDBF7CYT7/), [2019](https://www.surveymonkey.com/results/SM-S9JVNXNQV/) - number 2 answer to obstacle/frustrating/preventing question (losing by <10% to "Convincing coworkers/company/clients"); still rated most important area of Clojure to improve

When introducing people to programming in Clojure, long and complicated stacktraces are a leading complaint. Experienced Clojurians often don't feel it so acutely, but obscure stacktraces are a pain point in the Clojure experience.

Regarding Clojure tooling, CIDER is top dog. It is the most popular Clojure development environment, so a _lot_ of the bad experiences with Clojure errors happen in CIDER.

# New possibilities
For a long time, a better error message experience was impossible (or prohibitively difficult) due to limitations in the language itself: an error produced stacktraces with mostly Strings instead of data. As of Clojure 1.10, [this is no longer the case](https://clojure.org/reference/repl_and_main#_error_printing): Clojure now categorizes errors, returns information about the error as consistent `ex-data`, and provides functions like [`ex-triage`](https://clojuredocs.org/clojure.main/ex-triage), [`ex-str`](https://clojuredocs.org/clojure.main/ex-str), and [`Throwable->map`](https://clojure.github.io/clojure/clojure.core-api.html#clojure.core/Throwable-%3Emap).

The clojure.main REPL takes advantage of these, making many error messages extremely useful and concise. Alex Miller helpfully provided [a short list of examples](https://clojurians.slack.com/archives/C03S1KBA2/p1683143432992029?thread_ts=1683142572.115519&cid=C03S1KBA2):

 - `:::5`  - error during reading
 - `(let [x])` - error during macroexpansion caught by a macro spec
 - `(cond 1)` - error during macroexpansion thrown by the macro
 - `(defmulti 5 class)` - unexpected error during macroexpansion
 - `(def 5)` - compilation error
 - `(/ 1 0)` or (+ 1 :a) - evaluation error
 - `(deftype T [a])` `(defmethod print-method T [_ w] (throw (Exception. "boom")))` `(->T 1)` - printing error

The first of that list (`:::5`) doesn't look too bad in my CIDER, because the stacktrace isn't long and some of it is minimized. 

The second, `(let [x])`, looks like this in the Clojure CLI:

```
$ clj
Clojure 1.11.1
user=> (let [x])
Syntax error macroexpanding clojure.core/let at (REPL:1:1).
[x] - failed: even-number-of-forms? at: [:bindings] spec: :clojure.core.specs.alpha/bindings
```

Short and to the point. If I run the same in CIDER's REPL buffer, I get almost the same result...but that's not how I work. I almost never enter forms into the REPL buffer. Like many Clojurians, I prefer to work in an interactive style, sending forms from namespace buffers. With this, the CIDER error experience is not good:

```
  Show: Project-Only All 
  Hide: Clojure Java REPL Tooling Duplicates  (15 frames hidden)

2. Unhandled clojure.lang.Compiler$CompilerException
   Error compiling src/playground/error_examples.clj at (10:1)
   {:clojure.error/column 1,
    :clojure.error/line 10,
    :clojure.error/phase :macro-syntax-check,
    :clojure.error/source "/Users/daveliepmann/src/playground/playground.empty/src/playground/error_examples.clj",
    :clojure.error/symbol clojure.core/let}

             Compiler.java: 6972  clojure.lang.Compiler/checkSpecs
             Compiler.java: 6988  clojure.lang.Compiler/macroexpand1
             Compiler.java: 7075  clojure.lang.Compiler/macroexpand
             Compiler.java: 7161  clojure.lang.Compiler/eval
             Compiler.java: 7132  clojure.lang.Compiler/eval
                  core.clj: 3214  clojure.core/eval
                  core.clj: 3210  clojure.core/eval
    interruptible_eval.clj:   87  nrepl.middleware.interruptible-eval/evaluate/fn/fn
                  AFn.java:  152  clojure.lang.AFn/applyToHelper
                  AFn.java:  144  clojure.lang.AFn/applyTo
                  core.clj:  665  clojure.core/apply
                  core.clj: 1973  clojure.core/with-bindings*
                  core.clj: 1973  clojure.core/with-bindings*
               RestFn.java:  425  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   87  nrepl.middleware.interruptible-eval/evaluate/fn
                  main.clj:  437  clojure.main/repl/read-eval-print/fn
                  main.clj:  437  clojure.main/repl/read-eval-print
                  main.clj:  458  clojure.main/repl/fn
                  main.clj:  458  clojure.main/repl
                  main.clj:  368  clojure.main/repl
               RestFn.java: 1523  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   84  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:   56  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  152  nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
                  AFn.java:   22  clojure.lang.AFn/run
               session.clj:  218  nrepl.middleware.session/session-exec/main-loop/fn
               session.clj:  217  nrepl.middleware.session/session-exec/main-loop
                  AFn.java:   22  clojure.lang.AFn/run
               Thread.java:  829  java.lang.Thread/run

1. Caused by clojure.lang.ExceptionInfo
   Call to clojure.core/let did not conform to spec.
   {:clojure.spec.alpha/args ([x]),
    :clojure.spec.alpha/problems [{:in [0],
                                   :path [:bindings],
                                   :pred clojure.core.specs.alpha/even-number-of-forms?,
                                   :val [x],
                                   :via [:clojure.core.specs.alpha/bindings
                                         :clojure.core.specs.alpha/bindings]}],
    :clojure.spec.alpha/spec #<clojure.spec.alpha$regex_spec_impl$reify__2509@6c516aad>,
    :clojure.spec.alpha/value ([x])}

                 alpha.clj:  705  clojure.spec.alpha/macroexpand-check
                 alpha.clj:  697  clojure.spec.alpha/macroexpand-check
                  AFn.java:  156  clojure.lang.AFn/applyToHelper
                  AFn.java:  144  clojure.lang.AFn/applyTo
                  Var.java:  705  clojure.lang.Var/applyTo
             Compiler.java: 6970  clojure.lang.Compiler/checkSpecs
             Compiler.java: 6988  clojure.lang.Compiler/macroexpand1
             Compiler.java: 7075  clojure.lang.Compiler/macroexpand
             Compiler.java: 7161  clojure.lang.Compiler/eval
             Compiler.java: 7132  clojure.lang.Compiler/eval
                  core.clj: 3214  clojure.core/eval
                  core.clj: 3210  clojure.core/eval
    interruptible_eval.clj:   87  nrepl.middleware.interruptible-eval/evaluate/fn/fn
                  AFn.java:  152  clojure.lang.AFn/applyToHelper
                  AFn.java:  144  clojure.lang.AFn/applyTo
                  core.clj:  665  clojure.core/apply
                  core.clj: 1973  clojure.core/with-bindings*
                  core.clj: 1973  clojure.core/with-bindings*
               RestFn.java:  425  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   87  nrepl.middleware.interruptible-eval/evaluate/fn
                  main.clj:  437  clojure.main/repl/read-eval-print/fn
                  main.clj:  437  clojure.main/repl/read-eval-print
                  main.clj:  458  clojure.main/repl/fn
                  main.clj:  458  clojure.main/repl
                  main.clj:  368  clojure.main/repl
               RestFn.java: 1523  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   84  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:   56  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  152  nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
                  AFn.java:   22  clojure.lang.AFn/run
               session.clj:  218  nrepl.middleware.session/session-exec/main-loop/fn
               session.clj:  217  nrepl.middleware.session/session-exec/main-loop
                  AFn.java:   22  clojure.lang.AFn/run
               Thread.java:  829  java.lang.Thread/run
```

The relevant part doesn't even fit on my screen:

<img width="763" alt="Screenshot 2023-05-04 at 14 37 22" src="https://user-images.githubusercontent.com/974443/236206554-48b18a5c-b945-4365-a3dc-cdb375ddf25f.png">

How can we provide a nicer, more concise CIDER error UI?

**Comment by bbatsov** [comment]
There's a bit of backstory here https://github.com/clojure-emacs/cider/issues/2443 for whoever is interested. CIDER stacktrace functionality predates Clojure 1.10 by a few years and we never made a focused effort to update it once those changes landed for various reasons. 

In general I agree it'd better to support properly the extra info added in Clojure 1.10, so contributions in that direction would be most welcome. Back when it was introduced it was hard to parse Clojure (EDN) in Emacs, but that problem was solved by `parseedn`. At the very least it'd be nice if we make the error info maps collapsible, so they are not taking up so much screen estate.

**Reply by daveliepmann** [reply]
>There's a bit of backstory here #2443 for whoever is interested. 

Thanks, I didn't realize this had been on the back burner for close to 5 years.

> In general I agree it'd better to support properly the extra info added in Clojure 1.10, so contributions in that direction would be most welcome.

It's possible but unlikely that I'll be able to directly contribute a PR for this. But for folks who might be able to contribute, the first question I imagine would be to understand prior work. Does CIDER currently do anything at all with this data?

>At the very least it'd be nice if we make the error info maps collapsible, so they are not taking up so much screen estate.

I think I either don't understand or don't agree. We already have the ability to collapse stack frames, which works well. What's important – it seems to me, and I'd really like other folks to chime in here – is to de-emphasize the stacktrace entirely and present an **error summary** based on data, as Alex [suggested](https://github.com/clojure-emacs/cider/issues/2443#issuecomment-419655929) in 2018. 

Basically, the minimum is to match clojure.main's functionality, plus the ability to see the whole stacktrace. 

As it says in the [ticket](https://clojure.atlassian.net/browse/CLJ-2373) Alex linked: 

>  - exception messages should be small strings
>    - with the expectation that getMessage will be printed
>    - totally controlled by throwing code
>    - never print arbitrary data into a message
>  - flow data all the way to the edge
>    - ex-info & ex-data
>    - macroexpand spec errors have a well known format
>  - edge printing should be concise by default, configurable by user
>    - edge functions like repl-caught respect print and spec bindings
>    - concise summary by default
>    - all the details when you want

This is a longer-term concern, but the configuration described in that ticket feels like the way Magit has config options for various UI elements to be verbose, concise, collapsed, or hidden entirely. In their case it's whether to show the remote URL and "Untracked files", in CIDER we might want to do the same for things like error category and the standardized `ex-data`.

**Reply by bbatsov** [reply]
> I think I either don't understand or don't agree. We already have the ability to collapse stack frames, which works well. What's important – it seems to me, and I'd really like other folks to chime in here – is to de-emphasize the stacktrace entirely and present an error summary based on data, as Alex https://github.com/clojure-emacs/cider/issues/2443#issuecomment-419655929 in 2018.

I simply meant it'd be nice to collapse that huge error info map that started to appear in the stack frames in Clojure 1.10. Obviously it'd be even better if we parse it and use it to improve the error messages, but for me the map made the raw stacktraces more unwieldy as the frame that contains it looks huge. 

> This is a longer-term concern, but the configuration described in that ticket feels like the way Magit has config options for various UI elements to be verbose, concise, collapsed, or hidden entirely. In their case it's whether to show the remote URL and "Untracked files", in CIDER we might want to do the same for things like error category and the standardized ex-data.

Agreed. I assume for many people it'd be fine if they just see a short error summary in an overlay or the REPL buffer instead a dedicated stacktrace buffer.

**Comment by mk** [comment]
@daveliepmann thanks for opening this, I'd love to see things improved in cider as well. Somewhat related, I've noticed that if I augment and rethrow an Exception (using ex-cause) my cider will only show the cause, not the more useful original exception that's thrown:

<img width="1379" alt="Screen Shot 2023-05-04 at 17 18 02" src="https://user-images.githubusercontent.com/1187/236252523-e0b04982-76c0-453c-a9eb-3c8f2bcb0889.png">


If I collapse the cause and expand the original exception it looks good:
<img width="1379" alt="Screen Shot 2023-05-04 at 17 16 19" src="https://user-images.githubusercontent.com/1187/236252029-55cf7b7f-5ff2-4f33-aa55-d960f579e0f9.png">

~~Would this not be more what we want in general as the augmented (rethrown) exception has generally more context available? Otherwise why would you catch an rethrow?~~ After writing this and trying out more examples I'm noticing that in the case of `(let [x])` it's reversed, the cause is more useful than the rethrown `clojure.lang.Compiler$CompilerException`. So how do we decide which one to show?

**Comment by yuhan0** [comment]
Just pointing out that there's a configuration setting `cider-show-error-buffer` which might not be very discoverable - when set to nil, in addition to disabling the `*cider-stacktrace*` auto popup behaviour it displays the printed error message as an overlay at the end of the form.
https://docs.cider.mx/cider/usage/dealing_with_errors.html#configuration

Here's what that looks like when successively eval-ing forms from the buffer (note: `cider-eval-result-duration` set to `'change` so the overlays persist)
<img width="1067" alt="image" src="https://user-images.githubusercontent.com/22216124/236254135-100a8bbb-f785-4c60-b237-33049b2d1eed.png">

**Reply by yuhan0** [reply]
Perhaps the discoverability could be improved by renaming the defcustom to something like `cider-eval-error-style` with options `overlay` /  `stacktrace-popup` / `both`? 

The overlay style could arguably be the default, maybe with a `"See more in *cider-stacktrace* buffer"` tacked on to the end of the error message. 

I have a custom key to summon the stacktrace popup on demand as I find cider-selector's buffer switching confusing:
```
(defun +cider-goto-error-buffer ()
  (interactive)
  (pop-to-buffer cider-error-buffer))
```

**Reply by bbatsov** [reply]
Yeah, I think this might be a nice step in the right direction. Although, I guess it might become also a point of friction, given it's going to affect people quite visibly.

**Reply by jackrusher** [reply]
I'm only seeing the error message in the mini-buffer, not overlaid. Is there a separate setting to get the overlay behavior?

**Reply by daveliepmann** [reply]
If we go in this direction, yuhan0's `+cider-goto-error-buffer` seems like it would be need to be built-in. Being able to expand from concise error message to error buffer/stacktrace needs to be frictionless.

That said, I'm not sure I agree that this is the right solution. Aren't location of the error and behavior of opening a new buffer orthogonal concerns? I _like_ that an error pops up `*cider-error*`, it's just that I want that `*cider-error*` buffer to be arranged more clearly, with the short form of the error emphasized in some way, and providing ways to hide stacktraces by default and expand them with key commands. (Thinking longer-term, this would also be the place to put more UI over the `ex-info` data in the future.)

I don't think the clear, concise error message should only be available in the minibuffer or in an overlay which disappears when I start navigating.

**Reply by yuhan0** [reply]
@jackrusher perhaps you have `cider-use-overlays` set to nil? 

I didn't intend this post as a solution per se for the original discussion, and totally agree that the orthogonal problem of improving cider-error's UX / middleware error extraction needs closer attention.

In fact this feature came about from personal frustration that the popup buffer was so frequently unhelpful and verbose, that I would begin instinctively dismisssing the window as it appeared. This led to realising how more useful the concise error messages that were being emitted in my hidden repl buffer (exactly the point being made by the OP), and finding a way to surface these stderr messages without too much context switching. 

Now the act of bringing up the detailed stacktrace buffer is a deliberate decision ("expanding" is a good metaphor) instead of feeling like an annoying intrusion in those situations where I can just intuit the source of error ("agh, forgot to add :else to that last clause, go away `Unhandled clojure.lang.Compiler$CompilationException`") 

Related point re. overlays disappearing on navigation: this was another of my nits with eval results being too ephemeral, but when submitting a PR I don't feel licensed to affect others by changing the defaults? So many users probably never realise `cider-eval-result-duration` is an option they could set, bringing the experience up to parity (in my opinion) with other tools like Calva.

**Reply by bo-tato** [reply]
great tip, though it seems it also makes it so it doesn't show the stacktrace in cider debugger when the user explicitly requests it. I added to my config: `(advice-add 'evil-collection-cider-debug-stacktrace :after #'+cider-goto-error-buffer)`

**Comment by jackrusher** [comment]
@yuhan0 👍🏻 that was it, yeah. I wasn't thinking of this as a solution to the problem statement at the start of the thread, but I did want to try out that setup because I might prefer the ergonomics of a brief, persistent, in-place error message. :)

**Comment by daveliepmann** [comment]
I'm interested in revisiting this with some concrete solutions which are within easy reach. It's possible I have tunnel vision for the idea of an error summary so it would be good to see alternatives, or to flesh out the idea. Setting aside radical UI changes and existing workarounds using different configs: what is the simplest change which would address any of the examples from Alex's list? 

For instance, how easy/hard would it be to provide a summary in the existing `*cider-error*` buffer for "clojure.core/let did not conform to spec" ("because it didn't satisfy `even-number-of-forms?`"), and minimize the stacktrace?

**Reply by bbatsov** [reply]
It's a matter of updating the existing `stacktrace` middleware to actually process the extra metadata that got added in Clojure 1.10 and insert it somewhere in the stacktrace. I think that shouldn't be particularly hard. 

Recently much of the related functionality was extracted into https://github.com/clojure-emacs/haystack by @r0man, so perhaps he might be willing to help with this as well? 

For the "proper" solution - I guess we should have some alternative op that returns are "human-friendly" summary of the error instead of the stacktrace. And a some configuration options and wrapper commands for this in CIDER to select the behavior you want with respect to errors. That would require more work on the Emacs side of things (but relatively simple work) and probably a bit more work on the Clojure side of things as well.

**Reply by r0man** [reply]
Hello,

I had a quick look at the haystack analyzer using the `(let [x])` example and looked a bit around in the `*nrepl-messages*` buffer. 

From what I can tell, we already send data for the following keys to the client:

- `:clojure.error/line`
- `:clojure.error/column`
- `:clojure.error/phase`
- `:clojure.error/source`
- `:clojure.error/symbol`

Those keys and values are available in the `location` slot of NREPL message replies to the `analyze-last-stacktrace` op as pure data.

Looking at the code we should also have a `spec` slot describing the Clojure Spec error. However in the `let` example I did not find this slot in the message.

It looks we have a bug or outdated code and this `or` condition here [1] needs to be extended with:

```clojure
(or (:clojure.spec/problems data) 
    (:clojure.spec.alpha/problems data)
    (:clojure.spec/failure data)
    (:clojure.spec.alpha/failure data))
```

in order for the `spec` slot to be added. With this fix the analyzed cause will now have a spec slot which should contain the spec problems as data:

```
  :spec
  {:spec
   "#object[clojure.spec.alpha$regex_spec_impl$reify__2503 0x5e63fddf \"clojure.spec.alpha$regex_spec_impl$reify__2503@5e63fddf\"]",
   :value "([x])",
   :problems
   ({:in "[0]",
     :val "[x]",
     :predicate "even-number-of-forms?",
     :spec ":clojure.core.specs.alpha/bindings",
     :at "[:bindings]"})}
```

I think we have most things in place and it is mostly a matter of trying all those examples and making sure we provide all the necessary data in a usable form.

This is how I tried out the `let` example with the Haystack analyzer:

```
user> (let [x])
Syntax error macroexpanding clojure.core/let at (*cider-repl workspace/haystack:localhost:43425(clj)*:265:7).
[x] - failed: even-number-of-forms? at: [:bindings] spec: :clojure.core.specs.alpha/bindings

user> (clojure.pprint/pprint (haystack.analyzer/analyze *e))
```

I have 2 other NREPL projects going on at the moment, and will be on vacation soon, so I won't have time to work on this myself right now.

But I'm happy to assist anyone looking into it.

[1] https://github.com/clojure-emacs/haystack/blob/master/src/haystack/analyzer.clj#L337

**Comment by vemv** [comment]
I've created https://github.com/clojure-emacs/cider/issues/3418 which seems to me a very minimalistic and effective way of tackling 90% of the problem.

To be honest, I don't believe this typically is a "stacktrace" problem. A stacktrace is not the best medium of conveying `even-number-of-forms?` to the CIDER user.

The good news is, as seen in the https://github.com/clojure-emacs/cider/discussions/3338#discussioncomment-5807132 screenshot, CIDER already does a more than decent job at summarizing the issue and presenting it with a concise, overlay-based UI.

So, we just need to dynamically choose when, and when not to, display stacktraces. That's what #3418 is about.

Still, there are some nice to haves though:

* Stacktraces currently tend to include too many Clojure/tooling implementation details, even when hiding them via the `tooling` filter.
  * I've [already](https://github.com/clojure-emacs/haystack/pull/11) proposed a fix.
* I believe we can compile spec data  like `:pred clojure.core.specs.alpha/even-number-of-forms?` into a self-explanatory, user-friendly string
  * https://github.com/bhb/expound excels at this
  * Depending on the specific spec, this might be either have a low value add, or actually be very useful.
  * I don't plan to work on this, would be happy to discuss and review with any volunteer though.

**Comment by daveliepmann** [comment]
FWIW, here is the behavior in CIDER 1.8.2. 

Section `;;;` comments name the example; normal `;;` comments describe what showed up in my minibuffer (except where a stacktrace is indicated). Big steps forward :clap:

```clojure
(ns repl.cider-error-examples)

;;; error during reading
:::5

;; => Invalid token: :::5


;;; error during macroexpansion caught by a macro spec
(let [x])

;; => [x] - failed: even-number-of-forms? at: [:bindings] spec: :clojure.core.specs.alpha/bindings

;; CLI has an addition line (which seems helpful):
;;     Syntax error macroexpanding clojure.core/let at (REPL:1:1).
;;     [x] - failed: even-number-of-forms? at: [:bindings] spec: :clojure.core.specs.alpha/bindings


;;; error during macroexpansion thrown by the macro
(cond 1)

;; => cond requires an even number of forms

;; Again, CLI has an additional line:
;;     Syntax error macroexpanding cond at (REPL:1:1).    
;;     cond requires an even number of forms


;;; unexpected error during macroexpansion
(defmulti 5 class)

;; => Unexpected error (ClassCastException) macroexpanding defmulti at (src/repl/cider_error_examples.clj:31:1).
;; class java.lang.Long cannot be cast to class clojure.lang.IObj (java.lang.Long is in module java.base of loader 'bootstrap'; clojure.lang.IObj is in unnamed module of loader 'app')


;;; compilation error
(def 5)

;; The CLI's extra line feels less necessary in this situation, to me:
;;    Syntax error compiling def at (REPL:1:1).
;;    First argument to def must be a Symbol


;;; evaluation error
```
Note — these open a _\*cider-error\*_ buffer on my machine.

```clojure
(/ 1 0)

;; Unhandled java.lang.ArithmeticException
;;    Divide by zero

;;               Numbers.java:  190  clojure.lang.Numbers/divide
;;               Numbers.java: 3911  clojure.lang.Numbers/divide
;;                       REPL:   46  repl.cider-error-examples/eval19614

(+ 1 :a)

;; Unhandled java.lang.ClassCastException
;;    class clojure.lang.Keyword cannot be cast to class java.lang.Number
;;    (clojure.lang.Keyword is in unnamed module of loader 'app'; java.lang.Number
;;    is in module java.base of loader 'bootstrap')

;;               Numbers.java:  155  clojure.lang.Numbers/add
;;               Numbers.java: 3743  clojure.lang.Numbers/add
;;                       REPL:   58  repl.cider-error-examples/eval19616


;;; printing error
```
Another _\*cider-error\*_  and a stacktrace

```clojure
(deftype T [a])
(defmethod print-method T [_ w] (throw (Exception. "boom")))
(->T 1)
```
<img width="679" alt="Screenshot 2023-10-15 at 19 02 20" src="https://github.com/clojure-emacs/cider/assets/974443/791628a5-5edd-4a7e-b197-e622981faf55">

I wonder what other folks think of the "Syntax error macroexpanding clojure.core/let" lines which the CLI has but CIDER does not.

**Reply by vemv** [reply]
> I wonder what other folks think of the "Syntax error macroexpanding clojure.core/let" lines which the CLI has but CIDER does not.

This might be simply because CIDER doesn't always force scrolling so that all output is visible. This is what I see in the repl:

```
Syntax error macroexpanding clojure.core/let at (src/haystack/parser.cljc:30:1).
[1] - failed: even-number-of-forms? at: [:bindings] spec: :clojure.core.specs.alpha/bindings
user> 
```

...The first line indeed is not shown until you scroll to it.

> ;;; unexpected error during macroexpansion

This message could be handled a touch better, for which I've filed https://github.com/clojure-emacs/cider/issues/3521. Thanks!

> Note — these open a `*cider-error*` buffer on my machine.

Yes, that's expected behavior since those are runtime exceptions, not compilation/read/macroexpansion ones.

Our design/plan for them is as follows:

* For the vast majority of cases, honor `cider-show-error-buffer` which represents the user preference
* When a stacktrace is rendered, we now mark more frames as tooling ones, so that the noisy part can be toggled at will (following `defcustom cider-stacktrace-default-filters`)
* Certain runtime exceptions are compile-like, and we'll pretend that they are a compilation exception, applying the new logic
  * This was introduced [here](https://github.com/clojure-emacs/haystack/pull/14) and will be further developed [here](https://github.com/clojure-emacs/haystack/issues/15)

**Comment by daveliepmann** [comment]
> > I wonder what other folks think of the "Syntax error macroexpanding clojure.core/let" lines which the CLI has but CIDER does not.
> 
> This might be simply because CIDER doesn't always force scrolling so that all output is visible. This is what I see in the repl:
> 
> ```
> Syntax error macroexpanding clojure.core/let at (src/haystack/parser.cljc:30:1).
> [1] - failed: even-number-of-forms? at: [:bindings] spec: :clojure.core.specs.alpha/bindings
> user> 
> ```
> 
> ...The first line indeed is not shown until you scroll to it.

I don't think this is a scrolling issue. I see the same thing you do in the REPL:
<img width="755" alt="Screenshot 2023-10-15 at 20 54 32" src="https://github.com/clojure-emacs/cider/assets/974443/bbecbbcd-7379-4912-a7c4-db2ed2e61f2e">

That seems fine. What I believe is an issue is that like many CIDER users I almost never work in the REPL buffer, and am more concerned with behavior from a source buffer. I ran these examples from the `repl.cider-error-examples` ns in my "playground" project, and from there I'm pretty sure the first line is genuinely dropped. I don't see it in the overlay, nor the minibuffer, nor _\*Messages\*_:

<img width="1512" alt="Screenshot 2023-10-15 at 21 00 06" src="https://github.com/clojure-emacs/cider/assets/974443/4fae4e21-5759-436c-b297-8fd240216f00">

All the rest sounds excellent, thanks!

**Reply by vemv** [reply]
There's some intentional dropping, intended for the overlays, and that indeed affects the minibuffer (since it's considered equivalent to overlays, per `cider-use-overlays` logic). Was introduced in https://github.com/clojure-emacs/cider/pull/3493

I foresee that the vast majority of times `Unexpected error (ClassCastException) macroexpanding defmulti at (src/haystack/parser.cljc:21:1)` is redundant or useless info that steals precious screen real state.

(It's redundant because in the context of an interactive evaluation, the file and line generally always match the current ones)

As you hint, the full output is printed to the repl, but not to other places. It seems a fair compromise - even if you don't use the repl a lot, at least it's a 'log' (you probably can also use a dedicated buffer for output).

This concept is further expanded in today's https://github.com/clojure-emacs/cider/pull/3523

**Reply by daveliepmann** [reply]
> I foresee that the vast majority of times `Unexpected error (ClassCastException) macroexpanding defmulti at (src/haystack/parser.cljc:21:1)` is redundant or useless info that steals precious screen real state.

> (It's redundant because in the context of an interactive evaluation, the file and line generally always match the current ones)

Hmm. I see your point that the first line is somewhat verbose. Some words don't change across distinct error kinds.

At the same time, most of the expressions I evaluate are more than one line, so I see some worth in preserving line numbers. (And isn't this a concern in the context of #3495? I'm thinking of a scenario where a user with overlays turned off (so the error message only appears in the minibuffer) evaluates an entire buffer. Aren't line numbers vital in that case?)

I also see value in preserving the classification of the error — "Syntax error reading source" contains information distinguishing it from "Syntax error macroexpanding" and the others. One ready-to-hand example is `(defn 5)`, which gives these two lines:

```
Syntax error macroexpanding clojure.core/defn at (src/repl/cider_error_examples.clj:168:1).
5 - failed: simple-symbol? at: [:fn-name] spec: :clojure.core.specs.alpha/defn-args
```

I can figure out what happened given only the second line, but the first line points to the problem (`defn` itself) much more directly and with less jargon.

Or, consider `(fn)`:

<img width="613" alt="Screenshot 2023-10-15 at 22 26 09" src="https://github.com/clojure-emacs/cider/assets/974443/ea4da2f1-ebff-4b42-9139-42347d22efe7">

The information necessary to understand what went wrong and find it is precisely what has been removed: "Syntax error macroexpanding clojure.core/fn at (src/repl/cider_error_examples.clj:178:11)."

Lastly, the upside (preserving a single line of screen real estate) is not compelling to my preferences. Most of the time 1. I have only a small number of buffers open, 2. I'm on a large screen, and 3. if there's an error I want all the relevant info in one place.

Maybe this is worth letting folks configure according to their own tradeoffs?


> As you hint, the full output is printed to the repl, but not to other places. It seems a fair compromise - even if you don't use the repl a lot, at least it's a 'log' (you probably can also use a dedicated buffer for output).

I think I must respectfully disagree. To me, "the full output is printed to the repl" is not ideal, since I almost never look at the REPL buffer. My impression is that this is a fairly common CIDER workflow, and why not? Errors pop up, results go where I tell them to; the only info I can *only* get in the REPL buffer is a warning, which is by definition ignoreable.

**Reply by daveliepmann** [reply]
Another example. Say I fat-finger `C-k`:

<img width="559" alt="Screenshot 2023-10-15 at 22 43 58" src="https://github.com/clojure-emacs/cider/assets/974443/b9de5deb-fe66-4069-b610-f40733cb6750">

The hidden part of the error message seems pretty useful: "Syntax error macroexpanding letfn at (src/repl/cider_error_examples.clj:184:3)."

**Reply by vemv** [reply]
Alright! Fair enough, I've created https://github.com/clojure-emacs/cider/issues/3525

**Comment by vemv** [comment]
Today's round of input has been much useful, both in new info we've provided confirming the current/intended behavior starting from 1.8.0, and pointing out some necessary tweaks (many of which are already in transit).

Other than that, I'll lock this discussion simply b/c it's completed - I wouldn't want it to become a catch-all for future development.
